wgengine/wgtun/tun.go

```diff
@@ -0,0 +1,264 @@
+// Copyright (c) 2020 Tailscale Inc & AUTHORS All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+// Package wgtun provides a TUN struct implementing the tun.Device interface
+// with additional features as required by wgengine.
+package wgtun
+
+import (
+	"errors"
+	"io"
+	"os"
+	"sync"
+
+	"github.com/tailscale/wireguard-go/device"
+	"github.com/tailscale/wireguard-go/tun"
+	"tailscale.com/types/logger"
+	"tailscale.com/wgengine/filter"
+	"tailscale.com/wgengine/packet"
+)
+
+const (
+	readMaxSize   = device.MaxMessageSize
+	readOffset    = device.MessageTransportHeaderSize
+	MaxPacketSize = device.MaxContentSize
+)
+
+var (
+	ErrClosed       = errors.New("device closed")
+	ErrFiltered     = errors.New("packet dropped by filter")
+	ErrPacketTooBig = errors.New("packet too big")
+)
+
+// TUN wraps a tun.Device from wireguard-go,
+// augmenting it with filtering and packet injection.
+// All the added work happens in Read and Write:
+// the other methods delegate to the underlying tdev.
+type TUN struct {
+	logf logger.Logf
+	// tdev is the underlying TUN device.
+	tdev tun.Device
+
+	// buffer stores the oldest unconsumed packet from tdev.
+	// It is made a static buffer in order to avoid graticious allocation.
+	buffer [readMaxSize]byte
+	// bufferConsumed synchronizes access to buffer (shared by Read and poll).
+	bufferConsumed chan struct{}
+
+	// closed signals poll (by closing) when the device is closed.
+	closed chan struct{}
+	// errors is the error queue populated by poll.
+	errors chan error
+	// outbound is the queue by which packets leave the TUN device.
+	// The directions are relative to the network, not the device:
+	// inbound packets arrive via UDP and are written into the TUN device;
+	// outbound packets are read from the TUN device and sent out via UDP.
+	// This queue is needed because although inbound writes are synchronous,
+	// the other direction must wait on a Wireguard goroutine to poll it.
+	outbound chan []byte
+
+	// filterFlags control the verbosity of logging packet drops/accepts.
+	filterFlags filter.RunFlags
+	// filterMu locks the filter, since it may be in use
+	// during the netmap refresh when ipn/local updates it.
+	filterMu sync.Mutex
```

Comment:
I rarely recommend it, but in this case an atomic.Value (https://golang.org/pkg/sync/atomic/#Value) would probably be a good idea.

Just document it like:

    filter  atomic.Value // of *filter.Filter